hiclass/Explainer.py

```diff
@@ -0,0 +1,103 @@
+"""Explainer API for explaining predictions using shapley values."""
+import shap
+import numpy as np
+from copy import deepcopy
+from hiclass import (
+    LocalClassifierPerParentNode,
+    LocalClassifierPerNode,
+    LocalClassifierPerLevel,
+)
+
+
+class Explainer:
+    """Explainer class for returning shap values for each of the three hierarchical classifiers."""
+
+    def __init__(self, hierarchical_model, data=None, algorithm="auto", mode="linear"):
+        """
+        Initialize the SHAP explainer for a hierarchical model.
+
+        :param hierarchical_model: The hierarchical classification model to explain.
+        :param mode: Can be `tree`, `gradient`, `deep`, or `linear`
+        """
+        self.hierarchical_model = hierarchical_model
+        self.explainers = {}  # To store a SHAP explainer for each node
+        self.algorithm = algorithm
+        self.mode = mode
+        self.data = data
+
+        if mode == "tree":
+            self.explainer = shap.TreeExplainer
+        elif mode == "gradient":
+            self.explainer = shap.GradientExplainer
+        elif mode == "deep":
+            self.explainer = shap.DeepExplainer
+        elif mode == "linear":
+            self.explainer = shap.LinearExplainer
+        else:
+            raise ValueError(
+                f"Invalid mode: {mode}. Supported modes are 'tree', 'gradient', 'deep', and 'linear'."
+            )
+
+    def fit(self, background_data):
+        """
+        Fits SHAP explainers on the model for each node using background data.
+
+        :param background_data: Background data examples to initialize the SHAP values.
+                                This is often a sample of the training data.
+        """
+        # Assuming hierarchical_model.nodes provides access to individual node classifiers
+        for node in self.hierarchical_model.nodes:
+            model_at_node = self.hierarchical_model.get_model_at_node(node)
+            # Create a SHAP explainer for each node model
+            self.explainers[node] = shap.Explainer(model_at_node, background_data[node])
+
+    def explain(self, X):
+        """
+        Generate SHAP values for each node in the hierarchy for the given data.
+
+        :param X: Data for which to generate SHAP values.
+        :return: A dictionary of SHAP values for each node.
+        """
+        if isinstance(self.hierarchical_model, LocalClassifierPerParentNode):
+            return self._explain_lcppn(X)
+        elif isinstance(self.hierarchical_model, LocalClassifierPerLevel):
+            return self._explain_lcpl(X)
+        elif isinstance(self.hierarchical_model, LocalClassifierPerNode):
+            return self._explain_lcpl(X)
+        else:
+            raise ValueError(f"Invalid model: {self.hierarchical_model}.")
+
+    def _explain_lcppn(self, X):
```

Comment:
The method currently only works if X contains one sample. However, the naming X suggests it is callable for multiple samples. Could you maybe change the naming and/or write asserts to ensure you get what you want?